### PR TITLE
Add migration for restoring missing backoff jobs

### LINKED ISSUE #14329: Failed jobs are not reactivate after backoff
**Describe the bug**

[E2E tests](https://camunda.slack.com/archives/C013MEVQ4M9/p1694702959002779) failed because the send-ping job is not completed. All the instances that did not complete are stuck in `send-message-ping` or `send-message-pong` tasks.

image

Looking at the logs, publishing message failed due to resource exhausted (there were broker restarts and leader change around this time). So the job was failed with a retry backoff of 30 seconds. However, the worker logs and the exported records show that these jobs were not re-activated. 

Here is screenshot showing exported records in elastic search for some of these jobs.

tmp-78371-iIZGT5mhN4lx-image

tmp-78371-p0xwE2yFVwKD-image

tmp-78371-23tXS7l4N8jV-image

For a job that retried after backoff, you can see those records in ES. But we don't see them for the above jobs. Here is a an example of such a job were retrying after backoff worked successfully (from the same cluster and for same job type). 
tmp-78371-cK9yR3YYFP5d-image

Also ran `zbctl activate jobs "send-message-ping"` locally against this cluster. But it returned  no jobs, even though operate shows these jobs as not completed.

**To Reproduce**

Not sure. But I think restarts and leader change may have something to do with this.

**Expected behavior**

Failed jobs should be re-activatable after backoff timeout.

**Log/Stacktrace**

There were no relevant error in Zeebe. But here are the logs around 14.09.2023 16:33 CEST. All uncompleted tasks were created around 2023-09-14 16:32 - 16:33
[downloaded-logs-20230915-105930.zip](https://github.com/camunda/zeebe/files/12618554/downloaded-logs-20230915-105930.zip)

Here is the snapshot of partition 3 on which these jobs exist. 
[snapshot-partition-3.zip](https://github.com/camunda/zeebe/files/12617504/snapshot-partition-3.zip)

**Environment:**
- Zeebe Version: 8.1.16-SNAPSHOT (8.1.16-SNAPSHOT-stable-8.1-8522cb63)

### PR DESCRIPTION
**Pull request by nicpuppa**
## Description

This PR add a migration for restoring failed jobs that are deleted from JOB_BACKOFF column family. This is related to a bug resolved in a previous [PR](https://github.com/camunda/zeebe/pull/15419). This migration helps to restore older cluster affected by this.

## Related issues

closes #14329 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: engine/src/main/java/io/camunda/zeebe/engine/state/migration/JobBackoffRestoreMigration.java
Comment (Zelldon) [inline engine/src/main/java/io/camunda/zeebe/engine/state/migration/JobBackoffRestoreMigration.java:24]: ❌ I think we should make this a bit more clever, and not iterate over all failed jobs, it is enough to find one and directly return. So maybe a method which checks exist at least one failed job.

Comment (nicpuppa) [reply engine/src/main/java/io/camunda/zeebe/engine/state/migration/JobBackoffRestoreMigration.java:24]: In this way we can cover more cases, do you think is enough to run the migration if the backoff cf is empty and there is at least one failed job ? 
Iterate on the failed jobs ensure that the migration runs only if there are missing jobs in the backoff cf, is more accurate imho

Comment (koevskinikola) [reply engine/src/main/java/io/camunda/zeebe/engine/state/migration/JobBackoffRestoreMigration.java:24]: @Zelldon can you clarify a bit further what you mean with your comment. I agree with @nicpuppa that his approach is more robust.

Do you mean that we should stop iterating over the failed jobs if we detect only 1 failed job more than the number of jobs in the backoff cf?

**Review by koevskinikola — APPROVED** [review]
LGTM @nicpuppa :100: 

Let's clarify some of the feedback before merging the PR first.

**Comment by nicpuppa** [comment]
I changed a bit the behaviour of the migration, following @Zelldon suggestions. Let me know your thoughts @Zelldon @koevskinikola

File: engine/src/test/java/io/camunda/zeebe/engine/state/migration/JobBackoffRestoreMigrationTest.java
Comment (Zelldon) [inline engine/src/test/java/io/camunda/zeebe/engine/state/migration/JobBackoffRestoreMigrationTest.java:113]: ❓ 🤔 Isn't the test now the same as the above? MAybe I'm confused by the diff

File: engine/src/main/java/io/camunda/zeebe/engine/state/migration/JobBackoffRestoreMigration.java
Comment (Zelldon) [reply engine/src/main/java/io/camunda/zeebe/engine/state/migration/JobBackoffRestoreMigration.java:23]: Yes thanks or we just run it always, as there is no faster way to detect

Comment (nicpuppa) [reply engine/src/main/java/io/camunda/zeebe/engine/state/migration/JobBackoffRestoreMigration.java:23]: Since looking to just a failed job or run it always doesn't make too much difference, I will go with running the migration everytime, at least is a bit faster ⚡

File: engine/src/main/java/io/camunda/zeebe/engine/state/immutable/JobState.java
Comment (koevskinikola) [inline engine/src/main/java/io/camunda/zeebe/engine/state/immutable/JobState.java:40]: :wrench: Looks like we're not using this method anywhere, so we can remove it.

File: engine/src/test/java/io/camunda/zeebe/engine/state/migration/JobBackoffRestoreMigrationTest.java
Comment (koevskinikola) [inline engine/src/test/java/io/camunda/zeebe/engine/state/migration/JobBackoffRestoreMigrationTest.java:116]: :upside_down_face: Is the following a typo?
```suggestion
  public void shouldDoNothingIfFailedJobsAreTheSameAsBackoff() {
```

**Review by koevskinikola — APPROVED** [review]
Looks good @nicpuppa :tada: 

I have some small code improvement suggestions, but they are not impacting the correctness of the remaining code, so I'm approving the PR.

File: engine/src/test/java/io/camunda/zeebe/engine/state/migration/JobBackoffRestoreMigrationTest.java
Comment (nicpuppa) [reply engine/src/test/java/io/camunda/zeebe/engine/state/migration/JobBackoffRestoreMigrationTest.java:116]: yes 😄

**Comment by nicpuppa** [comment]
Thanks for the review @koevskinikola , I applied your suggestion. I will merge this PR after completing the CI

**Comment by nicpuppa** [comment]
❓ Do you know if we can create a performance test for this migration ? @koevskinikola 

The goal is to have some numbers to understand the impact of this addition. As this migration is supposed to run everytime could have an impact on the start time of the brokers. Also, this migration should be backported to previous versions ?

cc @Zelldon @abbasadel

**Comment by Zelldon** [comment]
@nicpuppa you could set up a similar performance test like we did for the engine and large state for example https://github.com/camunda/zeebe/blob/main/engine/src/test/java/io/camunda/zeebe/engine/perf/EngineLargeStatePerformanceTest.java

**Comment by nicpuppa** [comment]
I will have a look, thanks @Zelldon 👀

**Comment by koevskinikola** [comment]
@nicpuppa regarding backporting, I can see that there are multiple `version:8.x` labels set on the issue, so I think you'll need to backport the PR changes to `8.3.`, `8.2`, and `8.1`. But I would check with @abbasadel first, as these labels are new, and might have a different meaning now.

**Comment by nicpuppa** [comment]
Hey @Zelldon , I tried to create a performance test. But I might need an help on that. Let me know if you have a couple of minutes to review what I did so far and maybe give me some inputs 🙇

**Comment by Zelldon** [comment]
Hey @nicpuppa sorry I'm currently quite busy with other things. Maybe we can do it already async, do you have pushed it somewhere? Maybe on another branch?

**Comment by nicpuppa** [comment]
> Hey @nicpuppa sorry I'm currently quite busy with other things. Maybe we can do it already async, do you have pushed it somewhere? Maybe on another branch?

I can put this PR in draft and push here what I did so far. I got a NullPointer exception when try to invoke the migration, but I'm not able to figure out why

File: engine/src/test/java/io/camunda/zeebe/engine/perf/JobBackoffRestoreMigrationPerformanceTest.java
Comment (Zelldon) [inline engine/src/test/java/io/camunda/zeebe/engine/perf/JobBackoffRestoreMigrationPerformanceTest.java:114]: I don't think you need to reset the exporter here. You are not writing any records

File: engine/src/test/java/io/camunda/zeebe/engine/perf/JobBackoffRestoreMigrationPerformanceTest.java
Comment (Zelldon) [reply engine/src/test/java/io/camunda/zeebe/engine/perf/JobBackoffRestoreMigrationPerformanceTest.java:90]: I think you can remove the column families

File: engine/src/test/java/io/camunda/zeebe/engine/perf/JobBackoffRestoreMigrationPerformanceTest.java
Comment (Zelldon) [inline engine/src/test/java/io/camunda/zeebe/engine/perf/JobBackoffRestoreMigrationPerformanceTest.java:82]: Annotate this method with @Setup

File: engine/src/test/java/io/camunda/zeebe/engine/perf/JobBackoffRestoreMigrationPerformanceTest.java
Comment (nicpuppa) [reply engine/src/test/java/io/camunda/zeebe/engine/perf/JobBackoffRestoreMigrationPerformanceTest.java:114]: Yeah, is not the final test let me say. But the problem is that the test is not working, I'm able to create the state but then I got a NPE

File: engine/src/test/java/io/camunda/zeebe/engine/perf/JobBackoffRestoreMigrationPerformanceTest.java
Comment (Zelldon) [inline engine/src/test/java/io/camunda/zeebe/engine/perf/JobBackoffRestoreMigrationPerformanceTest.java:73]: This will be done by JMH testcase run if you annotate the method correctly

File: engine/src/test/java/io/camunda/zeebe/engine/perf/JobBackoffRestoreMigrationPerformanceTest.java
Comment (Zelldon) [inline engine/src/test/java/io/camunda/zeebe/engine/perf/JobBackoffRestoreMigrationPerformanceTest.java:61]: I don;t think that this parameter will be set ? 🤔 I think you need to use a field here

Comment (Zelldon) [reply engine/src/test/java/io/camunda/zeebe/engine/perf/JobBackoffRestoreMigrationPerformanceTest.java:61]: https://github.com/camunda/zeebe/blob/main/engine/src/test/java/io/camunda/zeebe/engine/perf/EngineLargeStatePerformanceTest.java

Comment (nicpuppa) [reply engine/src/test/java/io/camunda/zeebe/engine/perf/JobBackoffRestoreMigrationPerformanceTest.java:61]: I tried different setups for the test, but with the same result 😄 But I'm sure that I'm missing something, don't know what

Comment (Zelldon) [reply engine/src/test/java/io/camunda/zeebe/engine/perf/JobBackoffRestoreMigrationPerformanceTest.java:61]: You need to create the state more manually like the db and state classes, as I did in the linked benchmark class. Here we use the TestEngine, which is to create the engine. 

You need something similar, I\m not sure whether it will work with the extension.

**Comment by nicpuppa** [comment]
Hey @Zelldon , thanks to @oleschoenburg I found the issue with NPE. I clean the test. Now is ready to be properly reviewed.
Please have another look 👀 

cc @koevskinikola

**Comment by Zelldon** [comment]
Great that you got your test running. Sorry that I didn't come back to you earlier, I had to finish other stuff as well.

I'm wondering why we have this test now and what to do with it, especially with the results. Maybe you can provide some more insights, or expectations regarding the tests.


Let me briefly explain why we have the other tests, maybe it makes it a bit easier for you to reason about. 

We have recently added such JMH performance tests since we did several works to improve performance with respect to large states (creation of instances should be stable even when having larger state and snapshot replication and taking of snapshots should be fast even with larger state). To keep the performance, and not introduce any regressions we have these tests. These allow us to be notified if our performance goes under a certain limit, likely introduced due to a bug or refactoring we might do. Both tests verify the general performance of the system, with a large state.

Now coming to your newly added test, it looks to me like we are testing a quite niche part of the migration, which will only have an effect if we have certain jobs in state failed but not in the backoff column (edge-case). We are not testing the migration process in general or itself, which I would accept as something where we could write a goal like: "Migration should at max take 10 seconds, even on a larger data set". Here I'm not sure what the goal is we want to achieve or keep, so I don't see much value in the test itself. I hope this makes sense to you.

What I would have expected when we want to understand how the performance is impacted by this change (or I guess this was also meant by @abbasadel when asking [here](https://camunda.slack.com/archives/C037W9NMATG/p1701419826328279?thread_ts=1701330672.970579&cid=C037W9NMATG) please correct me if I'm wrong.) is the following:


We could do several [benchmarks](https://github.com/camunda/zeebe/tree/main/benchmarks):

1. Base-1: An upgrade from Version X to Version Y; Check how long the migration takes
2. Base-2: An upgrade from Version X to Version Y; With data in X, likely somehow data which has been corrupted (meaning the wrongly removed Jobs from the backoff column family; so using unfixed versions)
3. Fix-1: An upgrade from Version X to Version Y; Check how long the migration takes (without data; empty)
  4. Here we see the difference already to Base-1 
4. Fix-2:  An upgrade from Version X to Version Y; With data in X, likely somehow data which has been corrupted and can now be fixed with the new migration
  5. Here we should see the difference between Base-2

The question is still for me what to do with the results? Is it to improve the performance of this migration or just to do an assessment? Because other than that there is nothing we can avoid or can't do. We have to migrate regardless, to fix our current bug and avoid the job stuck in the limbo.

It is not like we had with multi-tenancy we were could have avoided migration due to a different design of the feature itself. 

So based on this I would just skip this benchmarking, remove the JMH test, and go ahead and merge the PR. 

Please @abbasadel (or @oleschoenburg @npepinpe) correct or even stop me if you see this completely differently.

**Comment by nicpuppa** [comment]
Thanks for raising this @Zelldon , and clarify why we introduce JMH performance tests in Zeebe. I can't disagree with you to be fair, after what you said. The new test just verify how long the migration take, but the result is useless.

My biggest concern was related to zeebe startup time, as this migration is running every time. But as you mention, there is no workaround, the backoff need to be restored.

I align with you on removing this test case and merge the PR 👍 

cc @abbasadel

**Comment by Zelldon** [comment]
> The new test just verify how long the migration take, but the result is useless.

I'm currently not sure whether this is true since the migration is not idempotent. Not sure but I think once JMH runs it once it might be maybe faking the performance as later iterations are faster as nothing has to be migrated again.

> My biggest concern was related to zeebe startup time, as this migration is running every time. But as you mention, there is no workaround, the backoff need to be restored.

I think that makes sense to me and I am also happy that you are interested in this. We can still do this, to at least verify that we are not breaking it and the restart time is reasonable. 

In general, we need to make sure that we know what we testing or want to verify AND that we can live with the consequences and results of such tests. Meaning if this means that migration is crap, we should find a better way how migration is done (Idk what this would mean), or at least write some details in the docs that timeouts have to be increased for the upgrade phase or something similar.

**Comment by Zelldon** [comment]
BTW I created https://github.com/camunda/zeebe/issues/15678 as follow-up improvement.

**Comment by nicpuppa** [comment]
> I'm currently not sure whether this is true since the migration is not idempotent. Not sure but I think once JMH runs it once it might be maybe faking the performance as later iterations are faster as nothing has to be migrated again.

The state should be recreated at every run, but I might be wrong

What is your suggestion here ? Creating a benchmark can help with the purpose ? But I'm not sure how to properly set it up.
I think it will be awesome if we can merge this before Friday

**Comment by abbasadel** [comment]
> there is nothing we can avoid or can't do. We have to migrate regardless, to fix our current bug and avoid the job stuck in the limbo.
>So based on this and to make progress, I would just skip this benchmarking, remove the JMH test, and go ahead and merge the PR.

Thanks for your input @Zelldon. I see your point that we need to fix the bug anyway and live with the side effect that the migration is not idempotent until we work on this [improvement](https://github.com/camunda/zeebe/issues/15678) 

Based on the assessment that @nicpuppa share with me that "the migration takes 4s to migrate 200000 jobs" which is reasonable "I think!"; let's go with your suggestion to skip the "skip this benchmarking, remove the JMH test, and go ahead and merge the PR"

**Comment by nicpuppa** [comment]
As mentioned by @abbasadel I reverted the addition of the performance test, tomorrow I will merge the PR 💪 🚀

**Comment by nicpuppa** [comment]
bors merge

**Pull request merged**
